src/Crypto/Secp256k1.hs

```diff
@@ -702,28 +745,30 @@ keyPairPubKeyXOTweakAdd KeyPair{..} Tweak{..} = unsafePerformIO . evalContT $ do
 
 -- | Compute a schnorr signature using a 'KeyPair'. The @ByteString@ must be 32 bytes long to get a @Just@ out of this
 -- function
-schnorrSign :: KeyPair -> ByteString -> Maybe Signature
+schnorrSign :: KeyPair -> ByteString -> Maybe SchnorrSignature
 schnorrSign KeyPair{..} bs
     | BS.length bs /= 32 = Nothing
     | otherwise = unsafePerformIO . evalContT $ do
         (msgHashPtr, _) <- ContT (unsafeUseByteString bs)
         keyPairPtr <- ContT (withForeignPtr keyPairFPtr)
         lift $ do
             sigBuf <- mallocBytes 64
-            -- TODO: provide randomness here instead of supplying a null pointer
-            ret <- Prim.schnorrsigSign ctx sigBuf msgHashPtr keyPairPtr nullPtr
-            if isSuccess ret
-                then Just . Signature <$> newForeignPtr finalizerFree sigBuf
-                else free sigBuf $> Nothing
+            gen <- newStdGen
```

Comment:
We can't do this. Signatures need to be deterministic or we need to make this function happen in IO. Otherwise various equality assumptions are going to be broken and it will be very difficult to debug.

The proper way to deal with this I think is to pass the StdGen in via a parameter. You can make it a Maybe if you want and use the deterministic process if it has `Nothing` passed in.

Perhaps also can you provide a justification for why you don't want to use the libsecp256k1 based deterministic nonce generation?

Reply:
Also please put the randomness change in as a separate commit.